### ISSUE TITLE
Bug: [no-useless-default-assignment] should report unnecessary default when field is present in both sides of ternary

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#11980]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=MYewdgzgLgBA3jAhjAvDA5AI0QL3TAX1RgFlEoALAOgCdEwATEAWwAoBKGAHhgAYqArDAD88JAC4MAMxAh8RSQkSSsiGvIDcQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tmUSWWS0AJogBmAQwR1xAygHMmAW0Qt0UAO7joTSODABfEPqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
const { a = 'baz' } = Math.random() < 0.5 ? { a: 'foo' } : { a: 'bar' };
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-useless-default-assignment": "warn"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

I expect the default on `a` to be reported, since it is provably unnecessary

### Actual Result

No linter report

### Additional Info

This was ostensibly introduced in https://github.com/typescript-eslint/typescript-eslint/pull/11908

cc @ulrichstark
